### PR TITLE
feat: check accessibilityLabelledBy in *ByLabelText queries

### PR DESCRIPTION
**Pull request by pafry7**
### Summary
Adds support for [accessibilityLabelledBy](https://reactnative.dev/docs/accessibility#accessibilitylabelledby-android) in `*ByLabelText` queries as in #1183.

### CODE REVIEW

File: src/queries/text.ts [hunk 1]
```diff
@@ -54,7 +54,7 @@ const getChildrenAsText = (children: React.ReactChild[]) => {
   return textContent;
 };
 
-const getNodeByText = (
+export const getNodeByText = (
```
Comment (pafry7) [inline src/queries/text.ts:57]: As it is used in two different queries, I thought about moving it to `helpers` folder.

File: src/queries/labelText.ts [hunk 2]
```diff
@@ -11,14 +14,38 @@ import type {
...
   instance: ReactTestInstance
 ): ((text: TextMatch) => Array<ReactTestInstance>) =>
   function queryAllByLabelTextFn(text) {
     return instance.findAll(
       (node) =>
         typeof node.type === 'string' &&
-        matchStringProp(node.props.accessibilityLabel, text)
+        (matchStringProp(node.props.accessibilityLabel, text) ||
+          matchAccessibilityLabelledByText(
+            instance,
```
Comment (pafry7) [inline src/queries/labelText.ts:45]: I don't like passing `instance` object here, but I couldn't figure out a better way, as I have to iterate over the whole tree to find an element with particular `nativeID`.

File: src/queries/text.ts (hunk 1 above)
Comment (mdjastrzebski) [reply src/queries/text.ts:57]: `helpers/matchers/text.ts` should be good

File: src/queries/labelText.ts [hunk 3]
```diff
@@ -11,14 +14,38 @@ import type {
...
+
 const queryAllByLabelText = (
   instance: ReactTestInstance
 ): ((text: TextMatch) => Array<ReactTestInstance>) =>
   function queryAllByLabelTextFn(text) {
     return instance.findAll(
       (node) =>
         typeof node.type === 'string' &&
-        matchStringProp(node.props.accessibilityLabel, text)
+        (matchStringProp(node.props.accessibilityLabel, text) ||
```
Comment (mdjastrzebski) [inline src/queries/labelText.ts:43]: Pls extract `matchAccessibilityLabel` function to encapsulate the `||` logic so that the condition here is simplier and contains only `&&` operatos?

File: src/queries/labelText.ts (hunk 2 above)
Comment (mdjastrzebski) [reply src/queries/labelText.ts:45]: I think it's fine as is. We have to find the matching element somehow, and that requires starting from the root.

File: website/docs/Queries.md [hunk 4]
```diff
@@ -195,7 +195,7 @@ getByLabelText(
 ): ReactTestInstance;
 ```
 
-Returns a `ReactTestInstance` with matching `accessibilityLabel` prop.
+Returns a `ReactTestInstance` with matching `accessibilityLabel` or text content of view referenced by `accessibilityLabelledBy`.
```
Comment (mdjastrzebski) [inline website/docs/Queries.md:198]:
```suggestion
Returns a `ReactTestInstance` with matching label:
* by matching `accessibilityLabel` prop
* or by matching text content of text view referenced by `accessibilityLabelledBy` prop
```

**Review by mdjastrzebski — CHANGES_REQUESTED** [review]
Good work @pafry7. The code looks pretty solid. I've left some comments regarding code organisation and docs.

File: src/queries/labelText.ts [hunk 5]
```diff
@@ -11,14 +14,38 @@ import type {
...
+) => {
+  if (!nativeID) {
+    return false;
+  }
+
+  return (
+    rootInstance.findAll(
+      (node) =>
+        filterNodeByType(node, Text) &&
+        matchStringProp(node.props.nativeID, nativeID) &&
```
Comment (AugustinLF) [inline src/queries/labelText.ts:30]: We should be able to match `<View nativeID="input"><Text>A label</Text></View>`, and I believe this implementation won't work.

Ideally we should do something like `rootInstance.findAll(node => matchStringProp(node.props.nativeID, nativeID)).findAll(node => filterNodeByType(node, Text) && getNodeByText(node, text))`.

And add the relevant test covering that.

File: src/queries/__tests__/labelText.test.tsx [hunk 6]
```diff
@@ -46,6 +50,12 @@ test('getByLabelText, queryByLabelText, findByLabelText', async () => {
     <Section />
   );
 
+  expect(getByLabelText(INPUT_LABEL).props.accessibilityLabelledBy).toEqual(
```
Comment (AugustinLF) [inline src/queries/__tests__/labelText.test.tsx:53]: Would it be possible to get that in a different `test()` case? I know there was only one so far, but I prefer explicitly named `test` cases for the different edge cases we want to cover. Makes it easier to figure out the issue when a test starts failing.

**Review by AugustinLF — CHANGES_REQUESTED** [review]

File: src/queries/text.ts (hunk 1 above)
Comment (pafry7) [reply src/queries/text.ts]: Hey, I've moved it to `helpers/matchers/matchTextContent`. Is it okay?

File: src/queries/labelText.ts (hunk 5 above)
Comment (pafry7) [reply src/queries/labelText.ts]: Hey, I've changed implementation and moved the test to a separate test case. Is it okay?

File: src/helpers/matchers/matchTextContent.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import React from 'react';
+import { Text } from 'react-native';
+import { ReactTestInstance } from 'react-test-renderer';
+import { matches, TextMatch } from '../../matches';
+import { TextMatchOptions } from '../../queries/text';
+import { filterNodeByType } from '../filterNodeByType';
+
+const getChildrenAsText = (children: React.ReactChild[]) => {
```
Comment (pafry7) [inline src/helpers/matchers/matchTextContent.ts:8]: `React.ReactChild` is deprecated. Do we want to change it right now?

**Comment by mdjastrzebski** [comment]
I've added `hacktoberfest-accept` label to mark this PR as eligible for Hacktoberfest contribution. I plan to review it again in following days.

File: src/queries/__tests__/labelText.test.tsx [hunk 8]
```diff
@@ -93,3 +104,13 @@ test('getAllByLabelText, queryAllByLabelText, findAllByLabelText', async () => {
...
+
+test('getByLabelText supports accessibilityLabelledBy', async () => {
+  const { getByLabelText, queryByLabelText } = render(<FormInput />);
+
+  expect(getByLabelText(INPUT_LABEL).props.accessibilityLabelledBy).toEqual(
+    INPUT_LABELLED_BY
+  );
+  const textInput = queryByLabelText(/input/g);
+  expect(textInput?.props.accessibilityLabelledBy).toEqual(INPUT_LABELLED_BY);
+});
```
Comment (mdjastrzebski) [inline src/queries/__tests__/labelText.test.tsx:116]: Some improvements to the test:
1. Test both direct and nested reference
2. Verify the matched element directly against `getByTestId` result

```suggestion
test('getByLabelText supports accessibilityLabelledBy', async () => {
  const { getByLabelText, queryByLabelText } = render(<>
    <Text nativeID={INPUT_LABELLED_BY}>{INPUT_LABEL}</Text>
    <TextInput testID="textInput" accessibilityLabelledBy={INPUT_LABELLED_BY} />
  </>);

  expect(getByLabelText(INPUT_LABEL)).toEqual(getByTestId('textInput'));
  expect(getByLabelText(/input/)).toEqual(getByTestId('textInput'));
});

test('getByLabelText supports nested accessibilityLabelledBy', async () => {
  const { getByLabelText, queryByLabelText } = render(<>
    <View nativeID={INPUT_LABELLED_BY}>
      <Text >{INPUT_LABEL}</Text>
    </View>
    <TextInput testID="textInput" accessibilityLabelledBy={INPUT_LABELLED_BY} />
  </>);

  expect(getByLabelText(INPUT_LABEL)).toEqual(getByTestId('textInput'));
  expect(getByLabelText(/input/)).toEqual(getByTestId('textInput'));
});
```

File: src/helpers/matchers/matchTextContent.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  return textContent;
+};
+
+/**
+ * Matches the given string property again string or regex matcher.
+ *
+ * @param prop - The string prop to match.
+ * @param matcher - The string or regex to match.
+ * @returns - Whether the string prop matches the given string or regex.
+ */
```
Comment (mdjastrzebski) [inline src/helpers/matchers/matchTextContent.ts:45]:
```suggestion
/**
 * Matches the given node's text content against string or regex matcher.
 *
 * @param node - Node which text content will be matched
 * @param matcher - The string or regex to match.
 * @returns - Whether the node's text content matches the given string or regex.
 */
```

**Review by mdjastrzebski — CHANGES_REQUESTED** [review]
@pafry7 Looks pretty solid. I've added some suggestions for improvements to code and tests.

File: src/queries/labelText.ts [hunk 10]
```diff
@@ -17,12 +18,48 @@ type ByLabelTextOptions = CommonQueryOptions & TextMatchOptions;
...
+) => {
+  if (!nativeID) {
+    return false;
+  }
+
+  const { exact, normalizer } = options;
+
+  return rootInstance
+    .findAll((node) =>
+      matches(nativeID, node.props?.nativeID, normalizer, exact)
```
Comment (pafry7) [inline src/queries/labelText.ts:41]: `React.Fragment` doesn't have neither `nativeID` prop nor `testID` and tests started to fail. I've extended the type of `text` in `matches` to `string | undefined`, because we are checking it anyway:
```
 if (typeof text !== 'string') {
    return false;
  }
```
Adding `typeof node.type === 'string'` would fix it, but then we have to add 
```
const isTextComponent = filterNodeByType(node, Text) | filterNodeByType(node, "Text")
```
to `matchTextContent`. 
I didn't know what fits best here. @mdjastrzebski

**Comment by mdjastrzebski** [comment]
Let's pause this waiting for #1222, as it has some overlapping changes, which might be useful in this PR.

**Comment by mdjastrzebski** [comment]
Rebased to current `main`.

**Comment by mdjastrzebski** [comment]
Rebased to `main`, cleaned up to account for recent code improvements in `main`.

@AugustinLF could you take another look at this PR?

**Review by mdjastrzebski — APPROVED** [review]
Looks good.

File: website/docs/Queries.md [hunk 11]
```diff
@@ -212,7 +212,9 @@ getByLabelText(
 ): ReactTestInstance;
 ```
 
-Returns a `ReactTestInstance` with matching `accessibilityLabel` prop.
+Returns a `ReactTestInstance` with matching label:
+- either by matching [`accessibilityLabel`](https://reactnative.dev/docs/accessibility#accessibilitylabel) prop
+- or by matching text content of view referenced by [`accessibilityLabelledBy`](https://reactnative.dev/docs/accessibility#accessibilitylabelledby-android) prop
```
Comment (MattAgn) [inline website/docs/Queries.md:217]: should we mention that this prop only works for android? it's also specified in the RN doc but I feel it's the first time an RNTL query would validate code working only on android so it might be worth pointing out

Comment (mdjastrzebski) [reply website/docs/Queries.md:217]: This is a good question, whether RNTL should work like `ios`, `android`, common part of `iOS` and `android` or either `iOS` or `Android`. So far we seem to take the last approach when it comes to `isHiddenFromAccessibility`: element is hidden if any iOS or Android prop indicates that it is hidden.

File: src/queries/text.ts [hunk 12]
```diff
@@ -19,7 +20,10 @@ import type {
   QueryByQuery,
 } from './makeQueries';
 import type { CommonQueryOptions } from './options';
-import { getHostComponentNames } from '../helpers/host-component-names';
+
+beforeEach(() => {
+  resetToDefaults();
+});
```
Comment (MattAgn) [inline src/queries/text.ts:26]: I don't understand these lines, this file is not a test file. Maybe an issue occurred during the rebase?

File: website/docs/Queries.md (hunk 11 above)
Comment (MattAgn) [reply website/docs/Queries.md:217]: true, did not notice it back when we implemented it. For isHiddenFromAccessibility, we could add a little note in [this page](https://callstack.github.io/react-native-testing-library/docs/api#ishiddenfromaccessibility)? Or is overkill?

**Review by MattAgn — APPROVED** [review]

File: website/docs/Queries.md (hunk 11 above)
Comment (mdjastrzebski) [reply website/docs/Queries.md:217]: It is something that would be good to analyse and discuss in the mid-future. We have various options here should figure out which one makes most sense from user POV. For now I would stick with the current approach of: "this works with either iOS or Android".

**Review by AugustinLF — APPROVED** [review]
